### ISSUE TITLE
fillDescriptions for PixelCPE classes

### ISSUE THREAD

**Issue by kpedro88** [cms-sw/cmssw#27660]
See https://github.com/cms-sw/cmssw/pull/27614#issuecomment-517022979:
> PixelCPEGenericESProducer and PixelCPETemplateRecoESProducer may benefit from fillDescriptions at some point soon to get rid of the `existsAs`

**Comment by kpedro88** [comment]
assign reconstruction

**Comment by mmusich** [comment]
So, I've started to have a look how to fix this, but I encountered an issue which exceeds my current knowledge of how to use `fillDescriptions`.
One of the parameters that are passed in the configuration is an `ESInputTag`

https://github.com/cms-sw/cmssw/blob/de3e59d268f593c359968493232218bcf97daa3e/RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py#L59

and as far as I can see in:

https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Types_Allowed_for_Arguments

that's not a supported type. 
Any suggestion on how to overcome this issue?
The way I came around it  is to:
1) change the name of `MagneticFieldRecord` to `MagneticFieldRecordInputTag` and make it a `cms.string("")`
2) create the `ESInputTag`on the fly from the `std::string` within the C++ source
3) set  `desc.setAllowAnything();` in order to ignore the now illegal parameter `MagneticFieldRecord` that is present in many pieces of configuration which I am not allowed to change (e.g HLT menus). 

This works, but looks somehow ugly / inefficient to me. Any suggestion is welcome.

**Comment by kpedro88** [comment]
@mmusich looks like the twiki is out of date .`cms.ESInputTag()` is allowed in our python configs.

**Comment by mmusich** [comment]
@kpedro88 than I must be doing something wrong.
If I use:
```C++
void PixelCPEGenericESProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
  edm::ParameterSetDescription desc;

  ....
 
  desc.add<edm::ESInputTag>("MagneticField",edm::ESInputTag(""));
 
  descriptions.add("PixelCPEGenericESProducer", desc);
}
```
it does not even compile. Also I could not find any matching pattern in CMSSW: https://cmssdt.cern.ch/dxr/CMSSW/search?q=desc.add%3Cedm%3A%3AESInputTag&case=true
What's the correct syntax?

**Comment by kpedro88** [comment]
@Dr15Jones @makortel can you comment on this? It's not clear to me why this would fail.

**Comment by makortel** [comment]
By quick look it seems to me that `fillDescriptions()` does not support `ESInputTag` (more specifically 
https://github.com/cms-sw/cmssw/blob/master/FWCore/ParameterSet/interface/ParameterDescriptionNode.h
)

@Dr15Jones I suppose we should add the support for `ESInputTag` and `vector<ESInputTag>`.

**Comment by Dr15Jones** [comment]
We should add ESInputTag support. I ‘m on vacation today so can’t do it right now.

**Comment by Dr15Jones** [comment]
@wddgit could you look into adding ESInputTag support to the ParameterSet description system?

**Comment by makortel** [comment]
@Dr15Jones @wddgit I actually started already and am almost finished.

**Comment by makortel** [comment]
> We should add ESInputTag support. I ‘m on vacation today so can’t do it right now.

Done here #27684.

**Comment by mmusich** [comment]
> > We should add ESInputTag support. I ‘m on vacation today so can’t do it right now.
> 
> Done here #27684.

Thanks @makortel

**Comment by mmusich** [comment]
I tried to address this issue here: https://github.com/cms-sw/cmssw/issues/27660

**Comment by perrotta** [comment]
+1
- Addressed by @mmusich in #27694, merged in CMSSW since CMSSW_11_0_0_pre7
